snake/snake_game_khr.py

```diff
@@ -0,0 +1,174 @@
+        self.symbol = symbol
+        self.body = [position]
+
+    def eat(self, position: tuple):
+        self.body.append(position)
+
+    def move(self, position: tuple):
+        self.body.append(position)
+        self.body.pop(0)
+
+    def choices(self):
+        next_positions = [
+            (1, 0),
+            (0, 1),
+            (-1, 0),
+            (0, -1)
+        ]
+        result = []
+        for next_position in next_positions:
+            new_positions = self.body[-1][0] + next_position[0], self.body[-1][1] + next_position[1]
+            for new_position in new_positions:
+                if new_position in self.body:
+                    continue
+            result.append(new_positions)
+        return result
+
+
+class Apple:
+    def __init__(self, symbol='$', position=(2, 2)):
+        self.position = position
+        self.symbol = symbol
+
+
+class GameOverError(Exception):
+    pass
+
+
+class Game:
+    def __init__(self, width, height):
+        self.symbol = "o"
+        self.width = width
+        self.height = height
+        self.board = Board(self.width, self.height)
+        self.snake = Snake()
+        self.apple = Apple()
+
+    def init_apple(self):
+        last_pos = self.apple.position
+        self.board = Board(self.width, self.height)
+        positions = []
+        for x in range(self.height):
+            for y in range(self.width):
+                z = (item for item in self.snake.body)
+                if (x, y) != last_pos and (x, y) not in z:
+                    if self.board.board[x][y] == ' ':
+                        positions.append((x, y))
+
+        if positions:
+            position = random.choices(positions)[0]
+            self.apple = Apple(position=position, symbol="$")
+            return self.apple.position
+        else:
+            raise GameOverError('No places for apple!')
+
+    def play(self):
+        game_cycles = 100
+        try:
+            self.render()
+            while game_cycles > 0:
+                i, j = self.snake.body[-1]
+                possible_moves = self.snake.choices()
+                next_moves = []
+                for move in possible_moves:
+                    if move in self.snake.body:
+                        continue
+                    if move[0] != 0 and move[1] != 0 and move[0] != (self.height-1) and move[1] != (self.width-1):
+                        next_moves.append(move)
+                # print(next_moves)
+                distances = []
+                for move in next_moves:
+                    dist = self.board.distance(move, self.apple.position)
+                    distances.append(dist)
+                # print(distances)
+                res = dict(zip(next_moves, distances))
+                try:
+                    min(res, key=res.get)
+                except ValueError:
+                    print('Game over!')
+                    break
+                next_move = min(res, key=res.get)
```

Comment:
```py
try:
    min(res, key=res.get)
except ValueError:
    print('Game over!')
    break
next_move = min(res, key=res.get)
```

is the same as 
```py
try:
   next_move = min(res, key=res.get)
except ValueError:
    print('Game over!')
    break
```
next_move will exist later in the code if there wont be exception (ValueError)

Reply:
I would even check if next_moves is not empty once you've collected it (line 128), otherwise raise an GameOverError right away.

Reply:
> I would even check if next_moves is not empty once you've collected it (line 128), otherwise raise an GameOverError right away.

So, that was the key reason why I tried to implement ValueError handling here. I got the following error each time when there where no min() of next_moves:
'VaueError: min() arg is an empty sequence'
So would it be better to handle it in _next_moves_ list section?